### DISCUSSION TITLE
Apex CPU time limit exceeded when using Calculate Rollup app

### DISCUSSION THREAD

**Discussion by nhardenVenn** [jamessimone/apex-rollup#345] [Q&A]
I've so far been unsuccessful in performing a full recalculate using the app.

I get this message: Apex CPU time limit exceeded in Apex jobs.

How can I troubleshoot this?

Many thanks,

Nate

**Comment by nhardenVenn** [comment]
One additional note: many of the rollup results were cleared out. Is that expected?

**Reply by jamessimone** [reply]
An unsuccessful calculation might result in the rollup values being cleared out. That's a side product of your first issue, in other words.

**Comment by jamessimone** [comment]
In order to address this issue, you'll likely want to update some values on the `Org Default` Rollup Control CMDT record that ships with Apex Rollup. Two values, in particular, may help:

* Max Lookup Rows Before Batching (I might lower this to 1000)
* Batch Chunk Size (I would try lowering this to 200; it defaults to 2000 which can lead to timeouts in orgs with a lot of other automation)

**Comment by nhardenVenn** [comment]
Thanks. Still getting the timeout error, though. I'll reduce the numbers further and see if that works. Is there anything else I can try?

**Reply by jamessimone** [reply]
I typically don’t hear about issues beyond tweaking those numbers when it comes to the full recalculate, and I might need to see some logs from the actual full recalc runs (which generates a number of Apex Debug logs, starting with a `/aura` log prior to proceeding with a number of batch logs).

If you haven’t already, make sure the `Is Rollup Logging Enabled` option on the Org default for Rollup Control is also enabled. I have, broadly speaking, two large dataset scenarios commonly being used within Apex Rollup:

* large number of children records
* large number of configured rollups
* sometimes both of those things together

if there’s some sort of issue which is still causing timeouts for your particular use-case, I’m eager to address that.

**Reply by staggerlee99** [reply]
Hi @jamessimone - first, thank you for creating and offering such an amazing tool! In the comment above you mention "large number of configured rollups." Question: How many is too many? We have 17 on one Object and I've been wrestling with the following error when running Apex-Rollup from a Scheduled Flow: "System.LimitException: Apex heap size too large: 10996724". The record count on the child Object is 5,887 and the parent object (Contact) has about 3700 records. I've lowered the Org Defaults to the numbers you suggested above.
Thanks in advance for any guidance!

**Reply by jamessimone** [reply]
@staggerlee99 I know of at least one person using many dozens of rollups on a single object. My understanding is that it took them some tweaking of the Rollup Control metadata to get everything working correctly. For your scheduled flow, how many records are part of your entrance criteria, and which rollup action are you calling? It might also be helpful to see the full stacktrace for where that limit exception is being thrown. The numbers you’re citing aren’t necessarily very large, so I’ll be curious to see the stacktrace to see where things are running south.

There are also some potential optimizations I can imagine for scheduled flows with large record volume, in particular, but I’ll have limited availability this week to take a look so I’ll have to take a look next week when I’m able. We can definitely work together to alleviate this issue for you.

**Reply by staggerlee99** [reply]
Hi @jamessimone thanks so much for your response. There are 1,011 records in the dataset. I'm using Count(8), SUM(4), Average(3),  and MAX(2). Each of these run succesfully when I start them individually from the "Recalculate Rollup" tab. I'll work on generating a Stack Trace - that's a new one for me :-). Thanks so much again!

**Reply by jamessimone** [reply]
@staggerlee99 I'm back "in the office" so to speak for this week. For the stack trace - if you have your Apex Developer Console open while running the full recalculate, a log should be generated and accessible in the log pane. Within the log that's generated, you should be able to search for `Exception` to see the full contents of the limit exception that was generated. Again, you're not operating under a large data set, so I'll be curious to see what the issue is. Hopefully we can coordinate this week to get more info and help address this for you.

**Reply by staggerlee99** [reply]
[apex-07L7V00005KZ1LoUAL.log](https://github.com/jamessimone/apex-rollup/files/9222186/apex-07L7V00005KZ1LoUAL.log)
Hi @jamessimone. Thx again for your response. So far, I'm seeing this issue: "JSON string exceeds heap size limit". Other messages: 
- 13:35:46:908 VARIABLE_ASSIGNMENT [129]|this.recordsToRollup|"List of size 3013 too large to display"|0x33c15701
The Log File is attached.
@jamessimone Hi James - any chance you have any time this week to take a look at this? Sorry for the imposition....but thanks in advance.

**Comment by nhardenVenn** [comment]
After adjusting the two settings you mentioned a second time, the full calculation ran without the CPU timeout error, but there are still 4,103 fund records that have blank results.

I'll work on getting the logs.

**Reply by jamessimone** [reply]
@nhardenVenn it's possible you were affected by #343, which I've just patched in the latest version (`v1.5.13`).
